node_tree.py

```diff
@@ -286,11 +291,59 @@ def timestamp(self):
     @property
     def has_link_count_changed(self):
         link_count = len(self.links)
-        if not link_count == self.tree_link_count: 
+        if not link_count == self.tree_link_count:
             # print('update event: link count changed', self.timestamp)
             self.tree_link_count = link_count
             return True
 
+    def fill_links_memory(self):
+        new_links = self.links.items()
+        self.sv_links[hash(self)] = self.links.items()
+        linked_sockets = []
+        output_sockets = []
+        for link in new_links:
+            linked_sockets.append((link[1].from_socket, link[1].to_socket))
+            output_sockets.append(link[1].from_socket)
+
+        self.sv_linked_sockets[hash(self)] = linked_sockets
+        self.sv_linked_output_sockets[hash(self)] = output_sockets
+    def use_link_memory(self):
+        if self.configuring_new_node:
+            # print('skipping global process during node init')
+            return
+        if self.is_frozen():
+            # print('not processing: because self/tree.is_frozen')
+            return
+
+        if not self.sv_process:
+            return
+        links_has_changed = self.sv_links[hash(self)] != self.links.items()
```

Comment:
This string can crash Blender easily:
https://docs.blender.org/api/current/bpy.types.Node.html#bpy.types.Node.insert_link

If you would like tracking of links creating/deleting you should look toward such events:
https://docs.blender.org/api/current/bpy.types.Node.html#bpy.types.Node.insert_link

First of all you should create list of possible ways of creating/deleting links. Then you should find an appropriate detection event for all ways. It can happen that Blender API does not allow tracking of links creation.

Reply:
I have changed the hash(self) to a tree_id method similar to the n_id

> This string can crash Blender easily:

Do you have any evidence on that? I have been playing with it and i haven't been able to crash Blender

Reply:
You can't compare them anyway. As I have said Blender objects does not have stable hash.

Reply:
well... it seems to work even after undo/redo

Reply:
instead of using the hash of the noderee, maybe simply use the the name of the nodetree, .. or whatever is currently used for the cache :)

renaming nodetree does pose a problem for that solution tho.

Reply:
https://yadi.sk/i/B_ugIFwT3LCZew

Reply:
@Durman you are right, that check fails after undo, it will return it has changed.
Then the function checks the linked_sockets memory and returns that none of the nodes has been affected so does not trigger the nodetree processing but updates the links_cache and keeps on with normality

Reply:
or it should.. Im working on it

Reply:
I have thought about how new links can be detected and came up with next solution. The idea is to generate addresses for each socket in a tree such as `node.n_id + socket.identifier` and create Sverchok links which will keep pare of socket addresses. Then if Sverchok link with current pair of socket addresses is not found it means new links is detected.

```Python
class SvLink(NamedTuple):
    from_socket_address: str
    to_socket_address: str

to_socket_sv_links = {to_socket_address: sv_link}

for bl_link in tree.links:
    to_socket_sv_link = to_socket_sv_links.get(address(bl_link), None)
    if to_socket_sv_link and to_socket_sv_link.is_equal(bl_link):
        continue
    else:
        new_link = bl_link
```